File: 2021/src/day2.rs
fn parse(input: &str) -> impl Iterator<Item = (&str, u64)> + '_ {
    input
        .lines()
        .map(|l| l.split_once(' ').unwrap())
        .map(|(dir, number)| (dir, number.parse().unwrap()))
}

pub fn part1(input: &str) -> u64 {
    let mut horizontal = 0;
    let mut depth = 0;

    for (dir, num) in parse(input) {
        match dir {
            "forward" => horizontal += num,
            "down" => depth += num,
            "up" => depth -= num,
            _ => panic!("invalid input!"),
        }
    }

    horizontal * depth
}

pub fn part2(input: &str) -> u64 {
    let mut aim = 0;
    let mut horizontal = 0;
    let mut depth = 0;

    for (dir, num) in parse(input) {
        match dir {
            "forward" => {
                horizontal += num;
                depth += aim * num;
            }
            "down" => aim += num,
            "up" => aim -= num,
            _ => panic!("invalid input!"),
        }
    }

    horizontal * depth
}

#[cfg(test)]
mod tests {
    use super::*;

    const EXAMPLE_INPUT: &'static str = "forward 5
down 5
forward 8
up 3
down 8
forward 2
";

    #[test]
    fn p1_example() {
        assert_eq!(part1(EXAMPLE_INPUT), 150);
    }

    #[test]
    fn p1_input() {
        let input = std::fs::read_to_string("input/day2").expect("reading input file");
        assert_eq!(part1(&input), 2070300);
    }

    #[test]
    fn p2_example() {
        assert_eq!(part2(EXAMPLE_INPUT), 900);
    }

    #[test]
    fn p2_input() {
        let input = std::fs::read_to_string("input/day2").expect("reading input file");
        assert_eq!(part2(&input), 2078985210);
    }
}

File: 2021-hs/src/Day2.hs
module Day2
    ( part1
    , part2
    , example
    ) where

example = "forward 5\n\
          \down 5\n\
          \forward 8\n\
          \up 3\n\
          \down 8\n\
          \forward 2"

parse input = map ((\[cmd, n] -> (cmd, read n :: Integer)) . words) $ lines input

part1 input = let (horiz, depth) = foldl apply_cmd (0, 0) (parse input) in horiz * depth
  where
    apply_cmd (horiz, depth) (cmd, n) = case cmd of
      "forward" -> (horiz + n, depth)
      "down" -> (horiz, depth + n)
      "up" -> (horiz, depth - n)
      _ -> undefined

part2 input = let (horiz, depth, aim) = foldl apply_cmd (0, 0, 0) (parse input) in horiz * depth
  where
    apply_cmd (horiz, depth, aim) (cmd, n) = case cmd of
      "forward" -> (horiz + n, depth + aim * n, aim)
      "down" -> (horiz, depth, aim + n)
      "up" -> (horiz, depth, aim - n)
      _ -> undefined
